### PR TITLE
Create Documentation For PluginRegistry

### LINKED ISSUE #386: Add documentation for how plugin registries are created
# Requested Update

### Description
With the addition of plugin registries (see #240), there is a need to add documentation to explain how these registries are created. It's a complex process involving discovering all available plugins in every package, and users should be informed of this process so they know the structure of how packages should be organized for the future.

It might be smart to add a new section GeoIPS documentation for just this process. 

### Background and Motivation
This is a large update for GeoIPS and should be documented appropriately. These plugin registries define how GeoIPS will locate the appropriate plugins, and is involved in many of the inner-workings of GeoIPS. It's important to have this process documented for both users and developers.

### PR DESCRIPTION
**Pull request by evrose54**
# Reviewer Checklist

* [x] Required ***existing tests*** pass (ie full_test.sh, others as appropriate)
* [x] NO REQUIRED ***unit tests*** (explain why not required)
* [x] NO REQUIRED ***integration tests*** (explain why not required)
* [x] Required ***documentation*** added for new/modified functionality
* [x] Required ***release notes*** added for new/modified functionality
* [x] NO REQUIRED ***updates to other repos*** (explain why not required)

# Related Issues
fixes #386 

# Testing Instructions
Run `./docs/build_docs.sh . geoips html_only` to produce the docs. Download the html directory from the build/sphinx folder, and place the index.html in your browser to review the changes.

# Summary
With the addition of create_plugin_registries and the new PluginRegistry Class, we have
a need to add documentation describing what these do, how they are used, and the
benefits that come from adding them to the main GeoIPS code. This branch does exactly
that, and adds a section under 'UserGuide' describing what these pieces of code actually
do.

    modified: geoips/docs/source/userguide/index.rst
    added: geops/docs/source/userguide/plugin_registries.rst

### CODE REVIEW

**Comment by evrose54** [comment]
Don't merge this until v1.12.0-release is merged.

**Comment by SRIKANTH-KUMAR** [comment]
This might not be very closely related to this part of the issue but definitely to the documentation. Adding here for discussion : 
There seems to be an issue with the PDF writing; the generated file is of 0 bytes. 

image

**Comment by biosafetylvl5** [comment]
✨ My first GeoIPS contribution - I copyedited what was written and fixed some spelling mistakes.

When I saw things that could be clearer I tried to take care of it myself, but there are two things I couldn't do without a better understanding of the plugin registries / geoips in general. 

The are:

1. The example is helpful, but needs to be more explicitly explained. Eg. walk the reader through the example "This line of the file ... because .. "
2. The benefits section has a sentence which is difficult to understand without context. The performance increases are partly due to the new code "only instantiating the plugin registry when a user requests a plugin", but the previous behavior isn't detailed which makes it difficult to understand why (explicitly) this is an improvement.
3. How does the plugin registry work? We are told why it's important, that it store info and some other details. But not *explicitly* what it does and how.

File: docs/source/userguide/plugin_registries.rst
Comment (evrose54) [inline docs/source/userguide/plugin_registries.rst:87]: Regarding your comment here. It might be useful to modify the lines above (84-85) to be reworded. We don't actually 'create' this entry in the plugin registry; that would take way too long to manually create each entry of the registry. Rather, when we run 'create_plugin_registries', these entries are generated. The user doesn't need to know how to create an entry, rather, they need to know how to create a valid plugin which will be added to the registry in the manner shown below when 'create_plugin_registries' is ran.

Comment (evrose54) [reply docs/source/userguide/plugin_registries.rst:87]: So, maybe that snippet of information could be something along the lines of "we've created the sector-plugin 'denver' as shown above. When 'create_plugin_registries' is ran, an entry representing sector-plugin 'denver' will be generated in the registry, following the structured path that is reflected by the plugin's interface type, and interface name. Such an entry can be seen below.'

Comment (biosafetylvl5) [reply docs/source/userguide/plugin_registries.rst:89]: Fixed!

File: docs/source/releases/v1_12_0.rst
Comment (biosafetylvl5) [inline docs/source/releases/v1_12_0.rst:544]: As long as this should be in the `v1_12_0` release notes, this looks like it's good to go to me!

**Comment by mindyls** [comment]
@mindyls look at this.

File: docs/source/releases/v1_12_0.rst
Comment (mindyls) [reply docs/source/releases/v1_12_0.rst:544]: I think this was added to 1.12.0 a while back, and has propagated into a number of PRs... This chunk of release note should be removed from 1.12.0 before merge.

File: docs/source/releases/v1_12_2a0.rst
Comment (mindyls) [inline docs/source/releases/v1_12_2a0.rst:36]: This should be removed also - from a merge conflict.

**Review by mindyls — APPROVED** [review]
Looks good to me.  @evrose54  @biosafetylvl5  note I removed the errant entry from 1.12.0 release note, and fixed some formatting in 1.12.2a0 release note (I left conversations on the changes to be resolved prior to merge)

**Comment by biosafetylvl5** [comment]
Thanks mindy! When exactly should we merge this PR? I'm still a bit foggy about how we manage revisions/releases

**Comment by mindyls** [comment]
> Thanks mindy! When exactly should we merge this PR? I'm still a bit foggy about how we manage revisions/releases

These can be merged now, once all the outstanding PRs are merged, I will start my release process (which will end up tagging 1.12.2a0 with all these merged PRs included).

As long as there is no vX.Y.Z VERSION RELEASE PR on a given repo (like this one: https://github.com/NRLMMD-GEOIPS/geoips_tropics/pull/2), then we are good to go to merge PRs.

File: docs/source/userguide/plugin_registries.rst
Comment (biosafetylvl5) [inline docs/source/userguide/plugin_registries.rst:115]: Thank you for making this easier to understand!

**Review by biosafetylvl5 — CHANGES_REQUESTED** [review]

File: docs/source/userguide/plugin_registries.rst
Comment (biosafetylvl5) [reply docs/source/userguide/plugin_registries.rst:152]: Fixed

File: docs/source/userguide/plugin_registries.rst
Comment (biosafetylvl5) [reply docs/source/userguide/plugin_registries.rst:130]: Fixed

File: docs/source/userguide/plugin_registries.rst
Comment (evrose54) [reply docs/source/userguide/plugin_registries.rst:56]: I used ``encapsulate`` here to express that these are the three essential plugin types supported by GeoIPS at the time being. I didn't want to limit us to those plugins type in case we were to add more, but we can always adjust the documentation in the future if we add more. Happy to change this to a different wording if you find it confusing.

Comment (biosafetylvl5) [reply docs/source/userguide/plugin_registries.rst:56]: I am a bit confused as to whether you're using `encapsulate` to mean the [OOP paradigm](https://www.geeksforgeeks.org/encapsulation-in-python/) or to generally mean something like "These keys represent the three plugin types ... ". Either is fine - it's just a bit ambiguous here.

Comment (evrose54) [reply docs/source/userguide/plugin_registries.rst:56]: I'll update that portion to be less ambiguous. You make a good point that people could be confused by that.

Comment (biosafetylvl5) [reply docs/source/userguide/plugin_registries.rst]: Thanks @evrose54 ! :)

**Review by biosafetylvl5 — APPROVED** [review]
LGTM! Just waiting to merge like mindy said above. Great job @evrose54 🎉

File: docs/source/releases/v1_12_2a0.rst
Comment (jsolbrig) [reply docs/source/releases/v1_12_2a0.rst]: This appears to be fixed.

**Comment by jsolbrig** [comment]
@evrose54 @biosafetylvl5 I just made a bunch of edits to this (sorry if I went overboard). Please take a look and make sure you're okay with the changes. If everything looks good, we can merge this.

I found a nice way to dynamically incorporate the help from `create_plugin_registries`. It will update if we update the command arguments. We can probably use the same method to document the CLI.

The resulting page looks like this now:

image

File: pyproject.toml
Comment (mindyls) [inline pyproject.toml:120]: Oooh, is this all that is needed to auto-document command line options?  Great.

File: docs/source/userguide/plugin_registries.rst
Comment (mindyls) [inline docs/source/userguide/plugin_registries.rst:63]: @jsolbrig This will be nice for the CLI.

File: docs/source/userguide/plugin_registries.rst
Comment (evrose54) [inline docs/source/userguide/plugin_registries.rst:1]: This all looks good to me. I really like the auto-generated documentation of help information for terminal commands as well. That's nice! One thing I think we should do for all documentation .rst files in the future is make all distribution statements a dropdown widget. Just looks better in my opinion. Thanks for going over this and tweaking it @jsolbrig. It looks great!

File: pyproject.toml
Comment (jsolbrig) [reply pyproject.toml:120]: Yep, this will work for the CLI too. I think we'd need to play with it a bit to get it to format nicely for all of the subcommands, but it is supposed to be able to do it.

**Pull request merged**
